dsl/flatten.cc

```diff
@@ -0,0 +1,352 @@
+#include "dsl/flatten.h"
+#include "ast/Helpers.h"
+#include "ast/ast.h"
+#include "ast/treemap/treemap.h"
+#include "core/core.h"
+
+#include <utility>
+
+using namespace std;
+
+// This DSL pass flattens nested methods, so that once we've reached a non-definition AST node (i.e. not a ClassDef or a
+// MethodDef) then we know that there are no MethodDefs lurking deeper in the tree. In order to work correctly, this
+// also needs to move some non-method-def things as well, specifically `sig`s and sends for method visibility
+// (e.g. `private` and the like), and it also updates the static-ness of some MethodDefs based on where they have
+// appeared in a nested context.
+//
+// So, a file like the following
+//
+// class A
+//   sig{void}
+//   private def foo
+//     sig{void}
+//     def self.bar; end
+//   end
+// end
+//
+// will morally be transformed into the following
+//
+// class A
+//   sig{void}
+//   private def foo; :bar; end
+//   sig{void}
+//   def bar; end   # notice the lack of `self.` here
+// end
+//
+// So no nested methods exist any longer, and additionally, the nested method `bar` has had the `self.` qualifier
+// removed: if you run the above code in Ruby, you'll find that `bar` is not defined as a class method on `A`, but
+// rather as a not-always-available instance method on `A`, so introducing it as a static method is not at all
+// correct. Finally, because methods evaluate to their corresponding symbols, the former location of `bar`'s definition
+// has been replaced with the verbatim symbol `:bar`.
+namespace sorbet::dsl {
+
+class FlattenWalk {
+    // This is what we keep on the stack: we need to know whether an item should be moved or not (i.e. whether it's
+    // nested or not) and keep a stack of the current 'staticness' level (i.e. how many levels of `def self.something`
+    // we're inside)
+    struct MethodData {
+        // If this is non-nullopt, then it means that we've allocated space for an expression and this is the index to
+        // that space. The reason we do this is so that we can keep the moved expressions in the same order we
+        // originally see them in: in the following example
```

Comment:
This seems fancy, do we need this?
Can we followup with removing this complexity?

Reply:
There are a few alternatives I could imagine, but we _do_ need some way of keeping track of the depth-first order at least for making sure that `sig`s are paired with their appropriate method definitions. I'm doing this in part because it's how the flatten logic is already implemented and it's reasonably easy to encapsulate (i.e. the actual definitions of things like `postTransformMethodDef` are pretty clean and the invariants here are maintained by putting all the logic in `ClassScope`) but I can imagine a few other approaches: maybe by building up the queue of things to move along with an index corresponding to its order and then sorting it before adding it to the end?

FWIW, this already does put everything at the end and only keeps the relative ordering of the stuff that's moved.